include/deal.II/lac/solver_cg.h

```diff
@@ -389,6 +462,481 @@ SolverCG<VectorType>::compute_eigs_and_cond(
+
+      IterationWorker(const MatrixType &        A,
+                      const PreconditionerType &preconditioner,
+                      const bool                flexible,
+                      VectorMemory<VectorType> &memory,
+                      VectorType &              x)
+        : IterationWorkerBase<VectorType, MatrixType, PreconditionerType>(
+            A,
+            preconditioner,
+            flexible,
+            memory,
+            x)
+      {}
+
+      void
+      do_iteration(const unsigned int iteration_index)
+      {
+        using Number = typename VectorType::value_type;
+
+        const auto operation_before_loop = [&](const unsigned int start_range,
+                                               const unsigned int end_range) {
+          Number *                       x = this->x.begin() + start_range;
+          Number *                       r = this->r.begin() + start_range;
+          Number *                       p = this->p.begin() + start_range;
+          Number *                       v = this->v.begin() + start_range;
+          std::array<Number, grain_size> prec_r;
+          if (iteration_index == 1)
+            {
+              for (unsigned int j = start_range; j < end_range; j += grain_size)
+                {
+                  const unsigned int length =
+                    std::min(grain_size, end_range - j);
+                  this->preconditioner.apply_to_subrange(j,
+                                                         length,
+                                                         r,
+                                                         prec_r.data());
+                  DEAL_II_OPENMP_SIMD_PRAGMA
+                  for (unsigned int i = 0; i < length; ++i)
+                    {
+                      p[i] = prec_r[i];
+                      v[i] = Number();
+                    }
+                  p += length;
+                  r += length;
+                  v += length;
+                }
+            }
+          else if (iteration_index % 2 == 0)
+            {
+              for (unsigned int j = start_range; j < end_range; j += grain_size)
+                {
+                  const unsigned int length =
+                    std::min(grain_size, end_range - j);
+                  DEAL_II_OPENMP_SIMD_PRAGMA
+                  for (unsigned int i = 0; i < length; ++i)
+                    r[i] -= this->alpha * v[i];
+                  this->preconditioner.apply_to_subrange(j,
+                                                         length,
+                                                         r,
+                                                         prec_r.data());
+                  DEAL_II_OPENMP_SIMD_PRAGMA
+                  for (unsigned int i = 0; i < length; ++i)
+                    {
+                      p[i] = this->beta * p[i] + prec_r[i];
+                      v[i] = Number();
+                    }
```

Comment:
Grain size, since one would like to merge these loops?

Reply:
One could merge these loops for `DiagonalMatrix`, yes. There is indeed some performance penalty to it. When I run https://github.com/kronbichler/mf_data_locality/blob/main/benchmark_precond/bench.cc and compared against the manually implemented CG method, I see a performance of 2.7 GDoF/s for degree 5 with deal.II's implementation, compared to 3.3 GDoF/s with the manual one. (These are on a 72 core node.) A considerable part of that is the fact that the reductions are done in non-vectorized form, and part is how I access the block-diagonal preconditioner. But I assume I would get 3.0-3.1 GDoF/s with those two eliminated, and the rest can only be obtained with the loop merging. The problem is that those parts are somewhat compute-limited, so every additional instruction adds to the run time, which is quite substantial here with several loops. However, I do not really see a better way, except letting the preconditioner give me a second interface with `apply(const unsigned int index, const Number value)` interface, which is clearly only possible for the diagonal or the identity operation. What do you think? Should we allow for two interfaces, one with the `apply_to_subrange` function and one with a `apply_to_entry`, or should we specialize internally for `DiagonalMatrix`?